RHCLOUD-38735 | feature: skip Cost Management applications in availability checks

MikelAlejoBR opened the pull request:
Enables being able to request sources from the Sources Monitor which are not empty —a source with just Cost Management applications is considered empty too— and skipping the availability status checks for Cost Management applications, to avoid doing the same work twice. More details in the Jira ticket.

## Jira ticket
[[RHCLOUD-38735]](https://issues.redhat.com/browse/RHCLOUD-38735)

MikelAlejoBR committed:
feature: skip Cost Management applications in availability checks

Enables being able to request sources from the Sources Monitor which are
not empty —a source with just Cost Management applications is considered
empty too— and skipping the availability status checks for Cost
Management applications, to avoid doing the same work twice. More
details in the Jira ticket.

RHCLOUD-38735

dehort commented on source_handlers_test.go:
```diff
@@ -1718,6 +1720,74 @@ func TestAvailabilityStatusCheck(t *testing.T) {
+		},
+	)
+
+	// Set the environment variables so that the URL to which we need to send the request gets properly built.
+	if err := os.Setenv("COST_MANAGEMENT_AVAILABILITY_CHECK_URL", "invalidurl"); err != nil {
+		t.Errorf("unable to set the URL for the Cost Management application: %err", err)
+	}
+	defer func() {
+		if err := os.Unsetenv("COST_MANAGEMENT_AVAILABILITY_CHECK_URL"); err != nil {
+			t.Errorf("unable to unset the URL for the Cost Management application: %err", err)
+		}
+	}()
```
Where does the COST_MANAGEMENT_AVAILABILITY_CHECK_URL env var actually get used?

It would be nice if we could modify this setting directly and then inject that setting (or object) into the thing we are testing, but that probably isn't possible at the moment.  (Something we can work on improving later on...)

MikelAlejoBR replied:
The environment variable gets generated here:

https://github.com/RedHatInsights/sources-api-go/blob/312de4d8016ce44e6077cf26a871c751b239e99a/model/application_type.go#L48-L69

Which then gets used in the following code:

https://github.com/RedHatInsights/sources-api-go/blob/312de4d8016ce44e6077cf26a871c751b239e99a/service/availability_check.go#L75-L86

This is why I needed to do this ugly thing to make the test work. I do agree that a refactor on that would be nice, so that we grab the environment variables when Sources starts and don't do the above env var dance all the time.

dehort commented on source_handlers_test.go:
```diff
@@ -1718,6 +1720,74 @@ func TestAvailabilityStatusCheck(t *testing.T) {
+		},
+	})
+
+	// Override the "mock source" so that we return the new mock with the source we just added above. Then, we need to
+	// make sure to revert it for the rest of the tests.
+	previousMockSourceDao := mockSourceDao
+	defer func() {
+		mockSourceDao = previousMockSourceDao
+	}()
+	mockSourceDao = &mocks.MockSourceDao{Sources: customSourceFixtures, RelatedSources: fixtures.TestSourceData}
```
I'm not sure I understand how this works.  What is using mockSourceDao?

MikelAlejoBR replied:
`mockSourceDao` is used by all the tests in the test suite. So if I just modify it by adding a new source to the fixtures, the tests which expect a very specific set of sources in the responses actually break.

dehort commented on internal_handlers.go:
```diff
@@ -42,9 +43,17 @@ func InternalSourceList(c echo.Context) error {
+	// here: https://issues.redhat.com/browse/RHCLOUD-38735.
+	var skipEmptySources = false
+	if skip := c.Request().Header.Get(headers.SkipEmptySources); skip != "" {
+		skipEmptySources = skip == "true"
```
Might be good to call `strings.ToLower()` on skip?

MikelAlejoBR replied:
Sounds good!

dehort commented on service/availability_check.go:
```diff
@@ -72,8 +72,14 @@ func RequestAvailabilityCheck(c echo.Context, source *m.Source, headers []kafka.
 	for _, app := range source.Applications {
+		// We do not want to send availability check requests for Cost Management applications, as detailed in
+		// https://issues.redhat.com/browse/RHCLOUD-38735.
+		if skipEmptySources && app.ApplicationType.Name == "/insights/platform/cost-management" {
```
Nit-picky...what about moving the `app.ApplicationType.Name == "/insights/platform/cost-management"` logic into a method?  Something like the following `isCostManagementApp(app)`?

```
func isCostManagementApp(app) bool {
  return app.ApplicationType.Name == "/insights/platform/cost-management"
}
```

Would that help communicate what that bit of logic is actually doing (kind of self-commenting) and hide the details about how it does (in case the logic changes in the future)?

MikelAlejoBR replied:
Done as well! Good idea!

dehort reviewed (approved):
I think this looks reasonable though I'm struggling to grok it all.  :)  I'm still learning sources...

Would it be more clear if the skip-empty flag was passed in as a query parameter instead of a header?  Just a though!  :thinking:

MikelAlejoBR committed:
refactor: lowercase the header's value

That way the "true" value can be written in any case.

RHCLOUD-38735

MikelAlejoBR committed:
refactor: extract application skipping logic into a function

Makes it easier to follow and to refactor if the logic changes in the
future.

RHCLOUD-38735

MikelAlejoBR committed:
fix: GCI

RHCLOUD-38735
